### PR TITLE
Decouple Brave Ads cmd-line arg flags

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/24201

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Confirm command-line args are working as expected for features, `staging=` and `debug=`

### CODE REVIEW

**Review by Brandon-T — APPROVED** [review]

File: vendor/bat-native-ads/src/bat/ads/internal/DEPS [hunk 1]
```diff
@@ -3,6 +3,9 @@ include_rules = [
 ]
 
 specific_include_rules = {
+  "did_override_variations_command_line_switches_parser_util.cc": [
+    "+components/variations/variations_switches.h",
+  ],
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/DEPS:8]: https://github.com/brave/brave-browser/issues/13793 will move `vendor/bat-native-ads` to `components/brave_ads` removing this layering violation. This issue is being prioritized.

**Review by goodov — APPROVED** [review]
`DEPS` lgtm

**Review by zenparsing — APPROVED** [review]

File: vendor/bat-native-ads/src/bat/ads/internal/account/utility/redeem_unblinded_payment_tokens/redeem_unblinded_payment_tokens.cc [hunk 2]
```diff
@@ -216,13 +216,8 @@ base::TimeDelta RedeemUnblindedPaymentTokens::CalculateTokenRedemptionDelay() {
...
   const base::Time now = base::Time::Now();
 
-  int64_t delay;
-
-  if (!g_is_debug) {
-    delay = kNextTokenRedemptionAfterSeconds;
-  } else {
-    delay = kDebugNextTokenRedemptionAfterSeconds;
-  }
+  int64_t delay = ShouldDebug() ? kDebugNextTokenRedemptionAfterSeconds
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/account/utility/redeem_unblinded_payment_tokens/redeem_unblinded_payment_tokens.cc:219]: nit: const int64_t delay

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/account/utility/redeem_unblinded_payment_tokens/redeem_unblinded_payment_tokens.cc]: Fixed

File: vendor/bat-native-ads/src/bat/ads/internal/base/unittest/unittest_command_line_switch_util.cc [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+}  // namespace
+
+void AppendCommandLineSwitches(
+    const CommandLineSwitchList& command_line_switches) {
+  if (command_line_switches.empty()) {
+    return;
+  }
+
+  CHECK(base::CommandLine::InitializedForCurrentProcess);
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/base/unittest/unittest_command_line_switch_util.cc:44]: Braces are missing. Should be:
CHECK(base::CommandLine::InitializedForCurrentProcess());

File: vendor/bat-native-ads/src/bat/ads/internal/flags/did_override_variations_command_line_switches/did_override_variations_command_line_switches_parser_util.cc [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+}  // namespace
+
+bool ParseVariationsCommandLineSwitches() {
+  if (HasSwitchWithValue(variations::switches::kFakeVariationsChannel) ||
+      HasSwitchWithValue(variations::switches::kVariationsOverrideCountry)) {
+    return true;
+  } else {
+    const base::flat_set<std::string> kCommandLineSwitches = {
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/flags/did_override_variations_command_line_switches/did_override_variations_command_line_switches_parser_util.cc:35]: As I understand, it is better to use a vector here, as we don't do a search through values.

File: vendor/bat-native-ads/src/bat/ads/internal/flags/environment/environment_command_line_switch_parser_util.cc [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+#include "bat/ads/internal/flags/environment/environment_command_line_switch_parser_util.h"
+
+#include <ostream>
+
+#include "base/notreached.h"
+#include "brave/components/brave_rewards/common/rewards_flags.h"
+
+namespace ads {
+
+// TODO(tmancey): Return absl::optional
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/flags/environment/environment_command_line_switch_parser_util.cc:15]: Has this todo became obsolete?

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/flags/environment/environment_command_line_switch_parser_util.cc]: Removed :-)

File: vendor/bat-native-ads/src/bat/ads/internal/flags/did_override_variations_command_line_switches/did_override_variations_command_line_switches_parser_util.cc (hunk 4 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/flags/did_override_variations_command_line_switches/did_override_variations_command_line_switches_parser_util.cc]: Fixed

File: vendor/bat-native-ads/src/bat/ads/internal/base/unittest/unittest_command_line_switch_util.cc (hunk 3 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/base/unittest/unittest_command_line_switch_util.cc]: Fixed

**Review by aseren — APPROVED** [review]

**Comment by bsclifton** [comment]
If uplifted, we'll also want to grab https://github.com/brave/brave-core/pull/14410

**Comment by tmancey** [comment]
NOTE: If uplifted, we should also grab https://github.com/brave/brave-core/pull/14411

**Pull request merged**